Add test that simulates 16k vault creations + transfers

KentShikama opened the pull request:
## Overview
<sup>_What this PR does, and why it's needed_</sup>

Note this is just for @JosDenmark @ArturGajowy and anyone else who is investigating the full genesis ceremony execution. This test is a simplified version of the full genesis ceremony execution that brings forth all of the main issues: how to get 16k vault creations + transfers done in a reasonable amount of time.

Instructions for reproducing flamegraphs

0. Start the test with `sbt casper/testOnly *RholangBuildTest*` (note there are two subtests so wait for the first to finish)
1. `jps` (to find PID - for me it is always named ForkMain) 
2. `./profiler.sh -d 100 -f /tmp/flamegraph.svg <PID>`

Flamegraphs

After commit 3420905 (see https://transfer.sh/N8nHM/flamegraph.svg) taken ~2 minutes after test starts. Around 1 vault creation/sec.

![Screen Shot 2019-07-24 at 2 07 38 AM](https://user-images.githubusercontent.com/6329898/61780714-dd9a2d00-adb7-11e9-9aa2-b97890f581eb.png)

After commit bbd296b (see https://transfer.sh/88oI1/flamegraph.svg) taken ~2 minutes after test starts. Around 5 vault creations/sec.

![Screen Shot 2019-07-24 at 2 19 07 AM](https://user-images.githubusercontent.com/6329898/61781634-8eed9280-adb9-11e9-8038-fec164d66b3d.png)

After commit bbd296b (see https://transfer.sh/broUv/flamegraph.svg) taken ~4 hour after test starts. Around 0.5 vault creations/sec.

![Screen Shot 2019-07-24 at 7 29 09 AM](https://user-images.githubusercontent.com/6329898/61802098-c6256900-ade4-11e9-9095-a824b0caa1bc.png)

After commit 1755631 (see https://transfer.sh/xU039/flamegraph.svg) taken ~10 minutes after test starts. Around 1.5 vault creations/sec.

![Screen Shot 2019-07-24 at 11 47 50 PM](https://user-images.githubusercontent.com/6329898/61852110-7f2f8600-ae6d-11e9-93ea-92571868168b.png)

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>



### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>



### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

KentShikama committed:
Remove roundTrip from getContinuations

This is safe because we do not modify the continuations
that are returned from getContinuations in the interpreter.

See https://rchain.atlassian.net/browse/RCHAIN-3257

KentShikama committed:
Use deterministic defaultGenesisPk for reproducibility

KentShikama committed:
Add test that simulates 16k vault creations + transfers

KentShikama committed:
Fix OOPE

ArturGajowy reviewed (changes requested):
There's no way we're merging this with the 300k lines of code that could be generated on the fly

KentShikama commented:
If it wasn't clear by the description, this is not meant to be merged

KentShikama committed:
Replace underlying List/HashMap of SortedParMap with TreeMap

This ensures that adding a new element into SortedParMap
is log(n)

KentShikama committed:
Use Base16.encode instead of toUtf8 for string scoring

Note if string is an invalid toUtf8 character, it will resolve to
the same value even if the strings are different.

KentShikama committed:
Memoize scoring of terms in SortedParMap

ArturGajowy commented on models/src/main/scala/coop/rchain/models/rholang/sorter/ExprSortMatcher.scala:
```diff
@@ -259,7 +260,7 @@ private[sorter] object ExprSortMatcher extends Sortable[Expr] {
       case gu: GUri    => ScoredTerm(e, Node(Score.URI, Leaf(gu.value))).pure[F]
       case GByteArray(ba) =>
-        ScoredTerm(e, Node(Score.EBYTEARR, Leaf(ba.toStringUtf8))).pure[F]
+        ScoredTerm(e, Node(Score.EBYTEARR, Leaf(Base16.encode(ba.toByteArray)))).pure[F]
```
base16.encode is a hellishly slow way to sort. We should rather use the byte array directly in the leaf, nad make sure that we sort Leaf(byteArray) using the array elements

ArturGajowy commented on models/src/main/scala/coop/rchain/models/SortedParMap.scala:
```diff
@@ -1,59 +1,63 @@
 
-  def apply(par: Par): Par = sortedMap(sort(par))
-
-  def contains(par: Par): Boolean = sortedMap.contains(sort(par))
```
as mentioned in discord, lookups would be more efficient if we kept both a hashmap and a treemap, and did the lookups inside the hashMap

KentShikama committed:
Add cache to SortedParMap Ordering[Par]

KentShikama committed:
Cache sort call in SortedParMap class

JosephDenman commented on models/src/main/scala/coop/rchain/models/SortedParMap.scala:
```diff
@@ -44,15 +44,25 @@ final case class SortedParMap(ps: TreeMap[Par, Par]) extends Iterable[(Par, Par)
 }
 
 object SortedParMap {
+  @SuppressWarnings(Array("org.wartremover.warts.Var", "org.wartremover.warts.Null"))
```
@ArturGajowy @KentShikama This is why I used `TreeMap[ScoredTerm[Par], Par]`. No cache, no resorting.

KentShikama committed:
Update genesis block execution test to start from wallets.txt

This allows for any size of wallets.txt to be test executed
